### PR TITLE
`<MultiSelect/>` - New callback API 

### PR DESCRIPTION
**Pull request by hepiyellow**
According to [this spec](https://github.com/wix/wix-style-react/pull/2697).
## `upgrade` prop

When `upgrade` prop is provided, then a new API will be used.

## What is NOT included yet in this PR

- `markFirstMatchingOption` is not implemented yet. The default behavior is `markFirstMatchingOption=false`
- `onSelect` is still called with old argument logic.
- No Documentation or Examples
- No Migration Doc

## DO NOT MERGE YET

Missing features and docs will be in separate PRs (to avoid a huge PR) which will have this branch as the target branch. 
When everything is ready, then we could merge to master.

### CODE REVIEW

File: src/MultiSelect/MultiSelect.js [hunk 1]
```diff
@@ -69,32 +77,43 @@ class MultiSelect extends InputWithOptions {
     this.setState({ pasteDetected: true });
   }
 
+  _splitValues(value) {
+    const delimitersRegexp = new RegExp(this.props.delimiters.join('|'), 'g');
+    const valueWithCommas = value.replace(delimitersRegexp, ',');
+    return valueWithCommas
+      .split(',')
```
Comment (yardnsm) [inline src/MultiSelect/MultiSelect.js]: You can pass a `RegExp` to `split`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/split#Parameters

File: src/MultiSelect/MultiSelect.js [hunk 2]
```diff
@@ -69,32 +77,43 @@ class MultiSelect extends InputWithOptions {
...
+      .map(str => str.trim())
+      .filter(str => str);
+  }
+
   _onChange(event) {
-    if (!this.state.pasteDetected) {
+    if (this.state.pasteDetected) {
+      if (this._isNewCallbackApi()) {
+        const value = event.target.value;
```
Comment (yardnsm) [inline src/MultiSelect/MultiSelect.js]: I've seen `event.value.target` used multiple times in this methods. Why not hoist it to the top?

File: src/MultiSelect/MultiSelect.js [hunk 3]
```diff
@@ -163,25 +197,39 @@ class MultiSelect extends InputWithOptions {
...
-      this.input.blur();
       return;
     }
 
-    if (this.props.onManuallyInput) {
-      this.props.onManuallyInput(
+    const { onManuallyInput, onTagsAdded } = this.props;
+    if (this._isNewCallbackApi()) {
+      const values = this._splitValues(inputValue);
+      onTagsAdded && values.length && onTagsAdded(values);
```
Comment (yardnsm) [inline src/MultiSelect/MultiSelect.js]: I'm not a fan of this chaining style in general, but you can remove it.

File: src/MultiSelect/MultiSelect.js [hunk 4]
```diff
@@ -163,25 +197,39 @@ class MultiSelect extends InputWithOptions {
     return tag ? { id, ...tag } : { id, label: value, theme };
   }
 
-  onSelect(options) {
+  onSelect(_options) {
```
Comment (yardnsm) [inline src/MultiSelect/MultiSelect.js]: Why `_options` and not keeping `options`?

File: src/MultiSelect/MultiSelect.js (hunk 1 above)
Comment (hepiyellow) [reply src/MultiSelect/MultiSelect.js]: ok. done.

File: src/MultiSelect/MultiSelect.js (hunk 4 above)
Comment (hepiyellow) [reply src/MultiSelect/MultiSelect.js]: I want to differentiate it from the `props.options`.
Maybe I should call the argument `selectedOptions`?
(which could also be misleading...)

**Comment by hepiyellow** [comment]
Approved by Shlomi.
Waiting for review on The Examples PR.
https://github.com/wix/wix-style-react/pull/2708

**Comment by hepiyellow** [comment]
Replaced by #2752

**Pull request closed without merging**